.github/workflows/sync-with-upstream.yaml

```diff
@@ -0,0 +1,81 @@
+# Create a PR to sync with the upstream template repo.
+# The template repo is https://github.com/jenstroeger/python-package-template.
+
+name: Sync with template repository
+on:
+  schedule:
+  - cron: 11 11 * * *
+
+permissions:
+  contents: read
+env:
+  # This is the username and email for the user who creates a branch and commits
+  # the changes. In an organisation that should be a dedicated devops account.
+  USER_NAME: jenstroeger
+  USER_EMAIL: jenstroeger@users.noreply.github.com
+
+jobs:
+  sync:
+    if: github.repository != 'jenstroeger/python-package-template'
+    runs-on: ubuntu-latest
+    permissions:
+      contents: write
+      pull-requests: write
+    steps:
+
+    - name: Check out template repository
+      uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # v3.3.0
+      with:
+        # If you decide to change the upstream template repository to a private one, uncomment
+        # the following argument to pass the required token to be able to check it out.
+        # token: ${{ secrets.<TEMPLATE_REPO_TOKEN> }}
+        repository: jenstroeger/python-package-template
+        fetch-depth: 0
+        path: template
+
+    - name: Check out current repository
+      uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # v3.3.0
+      with:
+        token: ${{ secrets.REPO_ACCESS_TOKEN }}
+        fetch-depth: 0
+        ref: main
+        path: repo
+
+    - name: Sync with template
+      env:
+        GH_TOKEN: ${{ secrets.REPO_ACCESS_TOKEN }}
+      run: |
+        LATEST_VERSION=$(cd template && git describe --tags --abbrev=0)
+        CURRENT_VERSION=$(test -f repo/.github/workflows/.template_version \
+          && cat repo/.github/workflows/.template_version || echo "v0.0.0")
+        echo "Latest version is $LATEST_VERSION and current version is $CURRENT_VERSION."
+
+        if [ $CURRENT_VERSION == $LATEST_VERSION ]; then
+          echo "Unable to find a new version, exiting..."
+        else
+
+          # Copy the files from the template to the current repo.
+          rsync --recursive --verbose --verbose --exclude tests/ --exclude src/ \
+            --exclude .git/ --exclude .github/workflows/.template_version template/ repo/
```

Comment:
I’ve used [`rsync`](https://linux.die.net/man/1/rsync) in the past but unfortunately I’m not deeply familiar with _all_ of its features.

My main concern about using `rsync` is that it clobbers the local files, and that then results in a larger change set and possible merge conflicts. I’ve been scratching my head about that for a while now, and wanted to share an alternative to the `rsync` approach:

- Check out the downstream repository as usual, and not into a subfolder;
- Add this upstream template repository as a remote ([docs](https://git-scm.com/docs/git-remote#Documentation/git-remote.txt-emaddem)):
    ```bash
    git remote add upstream https://github.com/jenstroeger/python-package-template.git
    ```
- Merge from the upstream repository and ask `git` to refer upstream changes ([docs](https://git-scm.com/docs/git-merge#Documentation/git-merge.txt-ort)):
    ```bash
    git merge --squash --allow-unrelated-histories --strategy ort --strategy-option theirs <REF>  # Where <REF> can be `main` or some other git ref.
    ```
    By using the [`theirs`](https://git-scm.com/docs/git-merge#Documentation/git-merge.txt-theirs) option for the default merge strategy `ort` we tell `git` to prefer the upstream changes.

Judging from a few local tests, this allowed me to merge the upstream changes into an independent private repository without merge conflicts; however, the resulting change set lost a few local changes, e.g. the package name of my repository whose merge conflict was resolved using the upstream template’s package name. But maybe that’s ok, because we need to review _and adjust_ the PR anyway 🤔

Reply:
Might be worth digging around [here](https://stackoverflow.com/questions/6557467/can-git-ignore-a-specific-line) or [here](https://stackoverflow.com/questions/16244969/how-to-tell-git-to-ignore-individual-lines-i-e-gitignore-for-specific-lines-of/16244970).

Also, regarding files/path we want to _exclude_ from the automatic merge (e.g. [here](https://stackoverflow.com/questions/6294355/git-how-to-exclude-files-from-merging)): that doesn’t really seem to be doable with `git` but perhaps a simple `git restore` after the above merge would do?

Reply:
These are interesting alternatives to explore, but I prefer to keep it as it is for now and incrementally improve.

Reply:
I created issue https://github.com/jenstroeger/python-package-template/issues/472 from the above comment.

Reply:
I also assigned that ticket to you, @behnazh.